File: day05/day5.hs
import Data.Char (toLower, toUpper)

main = do 
    input <- readFile "input.txt"
    putStrLn $ "Part 1: " ++ show (solvePart1 input)
    putStrLn $ "Part 2: " ++ show (solvePart2 input)

alphabet = "abcdefghijklmnopqrstuvwxyz"

reaction :: String -> String -> String
reaction stack [] = stack
reaction [] (x:xs) = reaction [x] xs
reaction (c:stack) (x:xs) 
    | c /= x  && toLower c == toLower x = reaction stack xs
    | otherwise = reaction (x:c:stack) xs

solvePart1 :: String -> Int
solvePart1 = (length . reaction [])

prepareInput:: String -> [String]
prepareInput input = map (const (reaction [] input)) alphabet

solvePart2 :: String -> Int
solvePart2 input = minimum . map length $ [reaction [] (filter ((/=c).toLower) input) | c <- alphabet]

File: day05/day5.py
def solve_part_1(string):
    stack = []
    for char in string:
        stack.append(char)
        if len(stack) > 1 and stack[-1] == stack[-2].swapcase():
            stack.pop()
            stack.pop()
    return len(stack)


def solve_part_2(string):
    best = None
    for char in 'abcdefghijklmnopqrstuvwxyz':
        to_replace = [char, char.capitalize()]
        s = [c for c in string if c not in to_replace]
        l = solve_part_1(s)
        if best is None or l < best:
            best = l
    return best


def main():
    with open('input.txt') as f:
        s = f.read()
    sol1 = solve_part_1(s)
    print('Part 1: {}'.format(sol1))
    sol2 = solve_part_2(s)
    print('Part 2: {}'.format(sol2))


if __name__ == "__main__":
    main()
